### PR TITLE
kvutils: fix validation of false negative key lookups

### LINKED ISSUE #5562: False negative key lookups by non-stakeholders in Sandbox
The committer validation in the Sandbox (new, not classic) and other KV Ledgers, checks contract key lookups using incorrect permissions, giving it the visibility of the submitter instead of the maintainers. That means if a non-maintainer has imperfect knowledge of the state of a contract key and submits and incorrect lookup, this will not be caught at that stage, and the transaction may be recorded as valid despite the incorrect lookup. 

The Participant will reject the transaction while reading from the ledger which means the Ledger API will behave correctly. However, since an incorrect transaction was recorded, there may be problems down the line:

1. If the transaction created other contracts with contract keys, those keys are now blocked, but the contracts blocking the keys are inaccessible
2. If the Sandbox is run with persistence, the persisted ledger now contains invalid transactions which need to be tidied up on migration

Since positive lookups have a liveness check, and always come from contract stakeholders, this problem only affects negative lookups coming from non-stakeholders. More concretely, it requires this constellation:

1. The ledger contains a contract `c` with key `k`, maintainers `m`.
2. `Alice` is not a stakeholder of `c` 
3. The maintainers `m` have delegated `Alice` the right to do a `lookupByKey k`.

In this case `Alice` can submit a transaction with with a negative lookup for `k`, which is committed. This is demonstrated by the below model. The scenario `Setup:initialize` fails with error 

```
Error: User abort: Submit failed with code 3: Reason: [InvalidLookup(com.daml.lf.transaction.Node$GlobalKey@a13ff73,None,Some(AbsoluteContractId(00544546bfafe538448b160cc987c2a8b1f0dcb5a2a9fb3daeec368049fcdd8d50)))]
```

This appears correct. Since `Alice` doesn't know anything about the `Delegated` contract with key `key`, she will have submitted an invalid transaction with a negative lookup. However, we can confirm that this transaction was in fact persisted because in the background we have created a contract with key `key2`. Any attempt by `Alice` to call the choice `CreateSecond` will fail due to a duplicate key error.

`Main.daml`:
```
module Main where

template Delegated
  with
    owner : Party
    k : Text
  where
    signatory owner
    key (owner, k) : (Party, Text)
    maintainer key._1

template Delegation
  with
    owner : Party
    delegate : Party
  where
    signatory owner
    observer delegate

    controller delegate can
      nonconsuming InvalidLookupAndCreateSecond: ()
        with
          p: Party
          expected: Optional (ContractId Delegated)
        do
          create Delegated with
            owner = owner
            k = "key2"
          lookupByKey @Delegated (p, "key")
          return ()

      nonconsuming CreateSecond: ()
        do
          create Delegated with
            owner = owner
            k = "key2"
          return ()
```

`Setup.daml`:
```
module Setup where

import Daml.Script

import Main

initialize : Script ()
initialize = do
  m <- allocatePartyWithHint "m" (PartyIdHint "m")
  alice <- allocatePartyWithHint "Alice" (PartyIdHint "Alice")

  delegated <- submit m do
    createCmd Delegated with
      owner = m
      k = "key"

  delegationCid <- submit m do
    createCmd Delegation with
      owner = m
      delegate = alice

  submit alice do
    exerciseCmd delegationCid InvalidLookupAndCreateSecond with
      p = m
      expected = None

  return ()
```

### PR DESCRIPTION
**Pull request by gerolf-da**
The current way of transaction validation and key lookup does not work
in the following scenario:
- the transaction coming from the participant has a negative key lookup
- the contract key state is loaded, which points to a contract that has
  been created by another transaction since the command interpretation
  on the participant
- the contract state for the contractId inside the contract key state
  has not been loaded, therefore the visibility and activeness check in
  `ProcessTransactionSubmission.lookupKey` returns `None`, even though
  there is in fact a contract with such a key.

To mitigate this issue, the `lookupKey` function must not check
activeness or visibility. If the submitter is not allowed to load the
contract via a key lookup, the authorization check should catch that.
Any other situation that is inconsistent should result in a rejection of
the transaction.

A small side effect of this change is that the `lookupContract` function
might try to load a contract that wasn't loaded from the kv state
earlier. This only happens because we first rebuild the entire
transaction before we compare it to the original transaction.
Previously we threw an exception in this case, but this results in a rather
ugly abort of the reinterpretation of the transaction. Therefore the
`lookupContract` function just returns `None`. Eventually the
transaction will be rejected during validation of the nodes, because
the `LookupByKey` nodes will be different.
This would not happen if we reinterpreted the transaction lazily and
compared the nodes as soon as they are produced with the corresponding
node of the original transaction.

Fixes #5562.
### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/test-common/src/main/daml/Test.daml [hunk 1]
```diff
@@ -469,20 +469,18 @@ template Delegation
...
 
     controller delegate can
       nonconsuming LookupByKeyDelegated: ()
         with
           p: Party
           k: Text
-          expected: Optional (ContractId Delegated)
         do
-          actual <- lookupByKey @Delegated (p, k)
-          assertMsg "lookup matches" (expected == actual)
```
Comment (fabiotudone-da) [inline ledger/test-common/src/main/daml/Test.daml:485]: Can you elaborate on the removal of these checks?

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/ActiveLedgerStateManager.scala [hunk 2]
```diff
@@ -112,7 +105,7 @@ class ActiveLedgerStateManager[ALS <: ActiveLedgerState[ALS]](initialState: => A
                   None
                 case None =>
                   // Contract not known
-                  Some(InactiveDependencyError(cid, predType))
+                  Some(Inconsistent(s"Could not lookup contract $cid"))
```
Comment (fabiotudone-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/store/ActiveLedgerStateManager.scala:108]: Is this an improvement related to the fix? I lack bg. on the previous error reporting but can you elaborate here?

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/ActiveLedgerStateManager.scala [hunk 3]
```diff
@@ -206,7 +199,8 @@ class ActiveLedgerStateManager[ALS <: ActiveLedgerState[ALS]](initialState: => A
                       )
                     } else {
                       ats.copy(
-                        errs = errs + InvalidLookup(gk, nlkup.result, currentResult)
+                        errs = errs + Disputed(
+                          s"Contract key lookup with different results: expected [${nlkup.result}], actual[${currentResult}]")
```
Comment (fabiotudone-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/store/ActiveLedgerStateManager.scala:203]: Is this an improvement related to the fix? I lack bg. on the previous error reporting but can you elaborate here?

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala [hunk 4]
```diff
@@ -271,15 +271,29 @@ private[kvutils] class ProcessTransactionSubmission(
...
+
+      // LookupByKey nodes themselves don't actually fetch the contract.
+      // Therefore we need to do an additional check on all contract keys
+      // that the referred contract satisfies the causal monotonicity invariant.
+      causalKeyMonotonicity = startingKeys.forall { key =>
+        val state = damlState(key)
+        val keyActiveAt =
+          Conversions.parseTimestamp(state.getContractKeyState.getActiveAt).toInstant
+        !keyActiveAt.isAfter(transactionEntry.ledgerEffectiveTime.toInstant)
+      }
```
Comment (fabiotudone-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala:291]: Could you shortly elaborate (in the comment perhaps) on the causal monotonicity definition? I read "all keys of contracts that are inputs of a transaction must NOT be active after the ledger effective time of that transaction". Is that correct and why is that (e.g. I'm thinking of non-consuming)?

Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala:291]: Also, I agree about correctness-first but I think it's also important to design with performance in mind. What's your take on that, i.e. how would you model the additional load required by the check and how did you plan for performance to be improved later (if at all needed / possible)?

File: ledger/test-common/src/main/daml/Test.daml (hunk 1 above)
Comment (gerolf-da) [reply ledger/test-common/src/main/daml/Test.daml:485]: This has to do with how a transaction is validated:
First it is fully interpreted, and only afterwards the original and the re-interpreted transaction are compared. In the case of a test with a false negative key lookup (the participant found no contract id for a given key, but the committer found one), this particular transaction was rejected due to this failed assertion, rather than a failed validation.
Ideally the validation would only produce nodes one by one and compare the produced node directly with the node in the original transaction. This would avoid doing additional work in a failure case. Unfortunately this is not how the validation is currently set up.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/ActiveLedgerStateManager.scala (hunk 2 above)
Comment (gerolf-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/store/ActiveLedgerStateManager.scala:108]: This is needed, because I had to fix the test to expect the correct error message due to the validation error. Without this change, the tests for sandbox-classic would not have passed, because sandbox-classic would produce a different error message.
So this is basically making kvutils based ledgers and sandbox-classic more consistent.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/ActiveLedgerStateManager.scala (hunk 3 above)
Comment (gerolf-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/store/ActiveLedgerStateManager.scala:203]: See my [previous comment](https://github.com/digital-asset/daml/pull/5608#discussion_r411343492).

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala (hunk 4 above)
Comment (gerolf-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala:291]: The `active_at` field depicts the `ledger_effective_time` of the transaction that created the contract with that particular key. The causal monotonicity requirement defines that all contracts that are an input to a transaction must be active at a ledger effective time <= the ledger effective time as the transaction under validation.

Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala:291]: > The `active_at` field depicts the `ledger_effective_time` of the transaction that created the contract with that particular key. The causal monotonicity requirement defines that all contracts that are an input to a transaction must be active at a ledger effective time <= the ledger effective time as the transaction under validation.

OK, why not calling it `ledger_effective_time`?

Comment (gerolf-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala:291]: This is in line with https://github.com/digital-asset/daml/blob/8e94c7af251eee175363b8b7f45e481a0aca90f7/ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto#L388-L390

Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala:291]: For clarity I'd still note that it's the `ledger_effective_time` of the transaction that created the contract, at least in the comment of both protos (but agreed, maybe not in this PR).

File: ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto [hunk 5]
```diff
@@ -438,6 +438,9 @@ message DamlContractKeyState {
   // The contract to which the key points to.
   // If unset the key is inactive.
   string contract_id = 1;
+
+  // The time from which the contract is active.
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto:442]: Please note here how this timestamp is determined (i.e., that it is ledger effective time) or rename it to `active_at_ledger_effective_time` or similar.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Version.scala [hunk 6]
```diff
@@ -80,6 +80,7 @@ object Version {
     *      * Add DamlSubmissionBatch message.
     *   4: * Remove application_id from DamlCommandDedupKey. Only submitter and commandId are used for deduplication.
     *      * Add deduplicatedUntil field to DamlCommandDedupValue to restrict the deduplication window.
+    *   5: * Add active_at to DamlContractKeyState to be able to check causal monotonicity for positive key lookups.
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Version.scala:83]: Please refer to the definition of causal monotonicity or explicitly spell it out like, e.g.:
```suggestion
    *   5: * Add active_at to DamlContractKeyState to be able to check causal monotonicity for positive key lookups, i.e., whether all contracts that are an input to a transaction are active at a ledger effective time <= the ledger effective time as the transaction under validation.
```

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala [hunk 7]
```diff
@@ -271,15 +271,29 @@ private[kvutils] class ProcessTransactionSubmission(
...
           recordTime,
           buildRejectionLogEntry(
             transactionEntry,
             RejectionReason.Disputed("DuplicateKey: Contract Key not unique")))
-    } yield r
+
+      // LookupByKey nodes themselves don't actually fetch the contract.
+      // Therefore we need to do an additional check on all contract keys
+      // that the referred contract satisfies the causal monotonicity invariant.
+      causalKeyMonotonicity = startingKeys.forall { key =>
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala:286]: To improve readability, can you move this into a helper function, please? This function is already too long.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala [hunk 8]
```diff
@@ -400,7 +414,12 @@ private[kvutils] class ProcessTransactionSubmission(
           key ->
             DamlStateValue.newBuilder
               .setContractKeyState(
-                DamlContractKeyState.newBuilder.setContractId(contractKeyState.fold("")(_.coid))
+                contractKeyState
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala:417]: To improve readability, please move this into a helper function like `updateContractKeyWithContractKeyState(...)`.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala [hunk 9]
```diff
@@ -426,11 +445,17 @@ private[kvutils] class ProcessTransactionSubmission(
...
+      // because it hinted at a missing input declaration. however, this only worked due to a bug
+      // in lookupKey.
+      //
+      // NOW: There is the possibility that the reinterpretation of the transaction yields a different
+      // result in a LookupByKey than the original transaction. This means that we might not have loaded
+      // the contract state data for the contractId pointed to by that contractKey.
+      // This is not a problem because after the transaction reinterpretation, we compare the original
+      // transaction with the reintrepreted one, and the LookupByKey node will not match.
+      // Additionally, all contract keys are checked to uphold causal monotonicity.
       if contractIsActiveAndVisibleToSubmitter(transactionEntry, contractState)
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala:459]: Please move the explanation to above the `contractState` definition -- the `if` clause should be right after the variable definition (otherwise the connection is not immediately clear between the filter clause and what it filters).

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala [hunk 10]
```diff
@@ -486,14 +511,20 @@ private[kvutils] class ProcessTransactionSubmission(
       .get(stateKey)
       .flatMap {
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala:512]: Can you remove this nested `.flatMap` logic, please? I.e., you could rewrite it with for comprehension for better readability.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala [hunk 11]
```diff
@@ -486,14 +511,20 @@ private[kvutils] class ProcessTransactionSubmission(
...
         _.flatMap { value =>
-          for {
-            contractId <- Some(value.getContractKeyState.getContractId)
-              .filter(_.nonEmpty)
-              .map(decodeContractId)
-            contractStateKey = contractIdToStateKey(contractId)
-            contractState <- inputState.get(contractStateKey).flatMap(_.map(_.getContractState))
-            if contractIsActiveAndVisibleToSubmitter(transactionEntry, contractState)
-          } yield contractId
+          // we cannot check whether the contract is active or not, because we might not have loaded it earlier.
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala:514]: The name of this function does not imply that we would be checking whether the contract is active or not, hence this explanation feels out of context. Is there a better place where this explanation could go?

File: ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto (hunk 5 above)
Comment (gerolf-da) [reply ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto:442]: I used the same name and comment as in `DamlContractState`. I can change it, but then these two fields are inconcistent.

Comment (miklos-da) [reply ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto:442]: Well, if this is a general concept (i.e., fields of type `Timestamp` general refer to a ledger effective time) then this does not need to be changed but we need to make sure we state such context in `daml_kvutils.proto`.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala (hunk 10 above)
Comment (gerolf-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala]: Fixed in 14a595f.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala (hunk 9 above)
Comment (gerolf-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala:488]: Fixed in 14a595f.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala (hunk 8 above)
Comment (gerolf-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala]: Fixed in 14a595f.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala (hunk 7 above)
Comment (gerolf-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala]: Fixed in 14a595f.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Version.scala (hunk 6 above)
Comment (gerolf-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Version.scala]: Fixed in 14a595f.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala [hunk 12]
```diff
@@ -426,11 +445,17 @@ private[kvutils] class ProcessTransactionSubmission(
     val stateKey = contractIdToStateKey(coid)
     for {
       // Fetch the state of the contract so that activeness and visibility can be checked.
-      contractState <- inputState.get(stateKey).flatMap(_.map(_.getContractState)).orElse {
-        logger.warn(
-          s"Lookup contract failed, contractId=$coid correlationId=${transactionEntry.commandId}")
-        throw Err.MissingInputState(stateKey)
-      }
+      contractState <- inputState.get(stateKey).flatMap(_.map(_.getContractState))
+      // PREVIOUSLY: we aborted the validation if the input state for the contract id wasn't loaded,
```
[the remark this answers is not in the document]
Comment (gerolf-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala]: Fixed in 14a595f.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/store/ActiveLedgerStateManager.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (gerolf-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/store/ActiveLedgerStateManager.scala]: Fixed in 14a595f.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala (hunk 4 above)
Comment (gerolf-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala:284]: I added comments to both fields in 14a595f.

File: ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto (hunk 5 above)
Comment (gerolf-da) [reply ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto:444]: I added more comments to both fields in 14a595f.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala (hunk 11 above)
Comment (gerolf-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala]: Unfortunately I don't have a better place. I think this is the best place for the time being.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/ContractKeys.scala [hunk 13]
```diff
@@ -47,18 +47,18 @@ final class ContractKeys(session: LedgerSession) extends LedgerTestSuite(session
...
+          .exercise(delegate, delegation.exerciseFetchByKeyDelegated(_, owner, key))
           .failed
 
         // lookup by key delegation is should fail during validation
         // Reason: During command interpretation, the lookup did not find anything due to privacy rules,
         // but validation determined that this result is wrong as the contract is there.
         lookupByKeyFailure <- beta
-          .exercise(delegate, delegation.exerciseLookupByKeyDelegated(_, owner, key, None))
+          .exercise(delegate, delegation.exerciseLookupByKeyDelegated(_, owner, key))
           .failed
```
Comment (simonmaxen-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/ContractKeys.scala:58]: This would seem to lead to unpredictable behaviour.  If lookup is done using lookupByKey on a non-existent key and it returns None I don't see why that behaviour should change if that key is then used by a different party in a way that is not visible to the original caller ?

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/ContractKeys.scala [hunk 14]
```diff
@@ -85,14 +85,14 @@ final class ContractKeys(session: LedgerSession) extends LedgerTestSuite(session
...
+          .exercise(delegate, delegation.exerciseFetchByKeyDelegated(_, owner, key))
           .failed
 
         // lookup by key should fail
         // Reason: During command interpretation, the lookup did not find anything due to privacy rules,
         // but validation determined that this result is wrong as the contract is there.
         lookupByKeyFailure <- beta
-          .exercise(delegate, delegation.exerciseLookupByKeyDelegated(_, owner, key, None))
+          .exercise(delegate, delegation.exerciseLookupByKeyDelegated(_, owner, key))
           .failed
```
Comment (simonmaxen-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/ContractKeys.scala:96]: See comment above.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/ContractKeys.scala (hunk 13 above)
Comment (gerolf-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/ContractKeys.scala:58]: Consider the following situation:
`Alice` can send a command that results in a transaction with a `NodeLookupByKey` with the right authorization, but `Alice` is not actually a stakeholder of the contract that the lookup would resolve to, therefore the `NodeLookupByKey` doesn't find a contract. Based on that fact, the DAML model can make a decision (i.e. create a contract with such a key). However, during validation the committer sees that there already exists such a contract. In that case the validation needs to fail, because the original transaction didn't have an accurate view of the ledger.

Similarly, if there is contention on the contract key, the validation shows that the original `NodeLookupByKey` and all following actions are invalid.

Comment (simonmaxen-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/ContractKeys.scala:58]: I see, so there are three possible outcomes.  Visible(conctractId), Available and Unavailable.  Could we change the API to reflect this more obviously ?

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala (hunk 11 above)
Comment (miklos-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committing/ProcessTransactionSubmission.scala]: OK
